Add first order texture historgram

blowekamp opened the pull request.

blowekamp committed:
ENH: Import TextureMovingHistogramImageFilter from old repository

This filter computes first order statistics (mean, min, max, variance,
sigma, skewness, kurtosis, and entropy), on a sliding window with and
efficient moving histogram based algorithm.

blowekamp committed:
ENH: Adding more documentation and cleaning code

Remove commented out code, fix indentation, fix style issue in test, etc..

blowekamp commented:
What do you think of the name of the classes? I think this name was derived from the classes used, and not purpose or output of the filter.

jbvimort commented:
If I can make a suggestion about the name of the filters: maybe we should try to fit to the names of the existing filters in ITK (particularly [here](https://github.com/InsightSoftwareConsortium/ITK/tree/master/Modules/Numerics/Statistics/include)) with itkScalarImageToCooccurrenceMatrixFilter, itkScalarImageToRunLengthFeaturesFilter... 

That's what I tried to do by naming the filters itkScalarImageTo[type of features]FeaturesImageFilter
So my suggestion would be to rename your filter itkScalarImageToFirstOrderTextureFeaturesImageFilter.

It might also be a good idea to rename itkScalarImageToTextureFeaturesImageFilter in itkScalarImageToSecondOrderTextureFeaturesImageFilter
. Even thought this modification would imply that we are not fitting anymore to the already existing filters in ITK named [itkScalarImageToTextureFeaturesFilter](https://itk.org/Doxygen/html/itkScalarImageToRunLengthFeaturesFilter_8h.html) (which is based on the same algorithms)

What do you think about of those suggestions?

blowekamp commented:
I think I disagree with the following the "To" convention for Image to Image of feature vector style filters.

The "To" conventions is generally used for base classes, and when a filter converts from one data type to another. There is itkImageToImageFilter, which is a base class where derived classes don't continue the "To" convention. Filters like "ScalarImageToCooccurrenceMatrixFilter", "HistogramToTextureFeaturesFilter", or "ImageToHistogram" do not produce an image as outputs but a change the data type or produce a set of features.

My expectation of an ITK filter name is that it maintains the same type ( even Image->VectorImage) unless it has the "To" naming conventions. As such I expect the convention of the simpler "[type of features]FeaturesImageFilter" to be followed for the filters in the module.

Comparing to the existing related filters, the propose convention has "FeaturesImageFilter" which indicates that the input and output will be images. "Features" shows it related to the existing filters, which just produce the scalars.

I think going with the simpler "FirstOrderTextureFeaturesImageFilters" should work, if you simplify the exist filter to remove the "ScalarImageTo".

Regarding "TextureFeatures"->"SecondOrderTextureFeatures", I'd think "ConcurrenceTextureFeatures" may be better, as it's a term that already exists in naming conventions. But there is the unfortunate existence of "TextureFeatures" alone refusing to these cooccurrence feature, however I think its important to add the additional specificity to distinguish and clarify which texture feature it is.

jbvimort commented:
Yes, it make a lot of sense! I agree with your ideas!

So the 3 filters would be named:
- FirstOrderTextureFeaturesImageFilter
- CoocurenceTextureFeaturesImageFilter
- RunLengthTextureFeaturesImageFilter

Does it look good to you?

blowekamp commented:
Agreed.

jbvimort commented:
How do you want to proceed? 
Will you change everything in this pull request? 
Or you prefer to push only the first order filter, and make the renaming part (of the existing filters) in a separated pull request? (that I will be glad to take care of, if needed)

blowekamp commented:
The renaming of the other classes should be a separate patch. Do you want to take it?

jbvimort commented:
Sure! I can take care of that! :smiley:

jbvimort commented:
I just finished: #14

blowekamp committed:
DOC: enumerate the output pixels components

blowekamp committed:
ENH: Add test from Gaussian distributed samples

Added new test, which uses Google Test, that runs the first order
feature filter on a generated Gaussian sample. This enables validation
of the results based on the expected values from a Gaussian sample. To
compute the tolerance, numpy was used to generate 10,000 Gaussian
distributions, where the distribution of the statistics where
examined. The expected entropy values is computed from expected 101^2
unique samples. A fix seed is used to ensure a repeatable distribution
is generated.

blowekamp commented:
Are the test compiled and run with the CI checks?

jbvimort commented:
No, it's still just building the Python Wheels for Linux, I need to work on the tests.

blowekamp committed:
ENH: Prefer explicit division of cached inverse.

The save inverse forces the precision to that of double, while
explicitly dividing may provide better numeric results.

blowekamp commented:
The filter is in pretty good shape now. Please let me know if any thing else needs to be improved.

jbvimort commented:
I'm sorry but it seems that I have an error when trying to build your branch:

 CMake Error at /home/jean-baptiste/Professional/Libraries/ITK/ITK/CMake/ExternalData.cmake:585 (message):
   Data file referenced by argument

DATA{/Input/cthead1.png}

does not lie under the top-level source directory

/home/jean-baptiste/test/ITKTextureFeatures

 Call Stack (most recent call first):
   /home/jean-baptiste/Professional/Libraries/ITK/ITK/CMake/ExternalData.cmake:485 (_ExternalData_arg)
   /home/jean-baptiste/Professional/Libraries/ITK/ITK/CMake/ExternalData.cmake:324 (ExternalData_expand_arguments)
   /home/jean-baptiste/Professional/Libraries/ITK/ITK/CMake/ITKModuleTest.cmake:127 (ExternalData_add_test)
   test/CMakeLists.txt:241 (itk_add_test)

Do you have a similar problem on your side?

blowekamp commented:
The actual line is this:
```
   DATA{${ITK_DATA_ROOT}/Input/cthead1.png}
```

I'd guess that "ITK_DATA_ROOT" is undefined when you are building. I am building with the module checkout in the ITK's "Modules/External" directory. Perhaps you are building it standalone against ITK?

blowekamp committed:
ENH: add cthead1.png data to local module

Referring to ITK_DATA_ROOT does not work when module is built outside
the ITK repository.

blowekamp commented:
I believe I address the Data issue, and I have rebased onto master to resolve the conflict.

jbvimort commented:
Thanks! :+1: 

I have another compilation error though:

Building CXX object test/CMakeFiles/TextureFeaturesGTestDriver.dir/itkFirstOrderTextureFeaturesImageFilterGTest.cxx.o
/home/jean-baptiste/test/ITKTextureFeatures/test/itkFirstOrderTextureFeaturesImageFilterGTest.cxx:23:25: fatal error: gtest/gtest.h: No such file or directory

blowekamp committed:
Rename filter to FirstOrderTextureFeaturesImageFilter

The utility histogram class is similarly renamed to
FirstOrderTextureHistogram.

blowekamp commented:
The addition of GoogleTest has happened recently in the ITK 4.12 development cycle. I am a fan of it for analytic test where you can compute and verify expected values.

I can understand not wanting to require ITK 4.12 for this module.

jbvimort commented:
Actually I'm using this remote module in a Slicer Extension that I'm working on right now [(BoneTextureExtension)](https://github.com/Kitware/BoneTextureExtension), and I'm concerned that GoogleTest  won't work properly with the version of ITK that Slicer is using right now?

jbvimort commented on include/itkFirstOrderTextureHistogram.h:
```diff
@@ -0,0 +1,141 @@
+      curCount += i->second;
+
+      const double p_x = double( i->second ) / count;
+      entropy += -p_x*std::log( p_x ) / itk::Math::ln2;
```
When itkTextureFeatures is build as an external module against ITK, itk::Math seems to be not detected, could it be useful to include itkMath.h ?

blowekamp committed:
COMP: Add missing itkMath  header

blowekamp committed:
COMP: address warnings about bracket initialization

Addresses warning of the following type:
itkFirstOrderTextureFeaturesImageFilterGTest.cxx:92:31: warning:
suggest braces around initialization of subobject [-Wmissing-braces]
  ImageType::IndexType idx = {r,r};

blowekamp committed:
ENH: Disable Google dependency and tests with ITK version <4.12

blowekamp commented:
I have disabled the GoogleTest dependency and tests with the version of ITK is <4.12, so this should help with compatibility with older ITK versions. For building with Slicer is testing enabled?

jbvimort commented:
The testing of ITK is disabled, and I should also disabled the testing of itkTestingFeatures in my Slicer's extension?
It looks like it is not going to be a problem for me?

jbvimort commented:
I also have an error as I'm building it as an external module of itk, it doesn't find GTest::GTest and GTest::Main (see error bellow).
Is Gtest correctly exported by ITK?


CMake Warning (dev) at /home/jean-baptiste/Professional/Libraries/ITK/ITK/CMake/ITKModuleTest.cmake:241 (add_executable):
   Policy CMP0028 is not set: Double colon in target name means ALIAS or
   IMPORTED target.  Run "cmake --help-policy CMP0028" for policy details.
   Use the cmake_policy command to set the policy and suppress this warning.

   Target "TextureFeaturesGTestDriver" links to target "GTest::GTest" but the
   target was not found.  Perhaps a find_package() call is missing for an
   IMPORTED target, or an ALIAS target is missing?
 Call Stack (most recent call first):
   test/CMakeLists.txt:259 (CreateGoogleTestDriver)

CMake Warning (dev) at /home/jean-baptiste/Professional/Libraries/ITK/ITK/CMake/ITKModuleTest.cmake:241 (add_executable):
   Policy CMP0028 is not set: Double colon in target name means ALIAS or
   IMPORTED target.  Run "cmake --help-policy CMP0028" for policy details.
   Use the cmake_policy command to set the policy and suppress this warning.

   Target "TextureFeaturesGTestDriver" links to target "GTest::Main" but the
   target was not found.  Perhaps a find_package() call is missing for an
   IMPORTED target, or an ALIAS target is missing?
 Call Stack (most recent call first):
   test/CMakeLists.txt:259 (CreateGoogleTestDriver)

blowekamp commented:
Thanks for trying it again locally. I will build the module outside of ITK to try to reproduce the issue.

blowekamp commented:
I believe that patch should fix the issue:
http://review.source.kitware.com/#/c/22478/

jbvimort commented:
Everything looks perfect to me! :+1: 
Thank you for all the work! :smiley:
